File: 2006/J064.pas
var
  a: array[1..500, 0..500] of longint;
  b, covered: array[1..500] of boolean;
  i, j, n: longint;
begin
  for i := 1 to 500 do
  begin
    covered[i] := false;
    b[i] := false
  end;
  read(n);
  for i := 1 to n do
  begin
    read(a[i][0]);
    for j := 1 to a[i][0] do
      read(a[i][j])
  end;
  for i := n downto 1 do
  begin
    if (not covered[i]) then
    begin
      b[i] := true;
      for j := 1 to a[i][0] do
      begin
        if (covered[a[i][j]]) then
        begin
          writeln(-1);
          halt
        end;
        covered[a[i][j]] := true
      end
    end
  end;
  for i := 1 to n do
    if (b[i]) then
      writeln(i)
end.

File: 2006/J064.cpp
#include <cstdio>
int a[501][501];
bool covered[501], b[501];
int main() {
  int n;
  scanf("%d", &n);
  for (int i = 1; i <= n; i++) {
    scanf("%d", &a[i][0]);
    for (int j = 1; j <= a[i][0]; j++) {
      scanf("%d", &a[i][j]);
    }
  }
  for (int i = n; i >= 1; i--) {
    if (!covered[i]) {
      b[i] = true;
      for (int j = 1; j <= a[i][0]; j++) {
        if (covered[a[i][j]]) {
          printf("-1\n");
          return 0;
        }
        covered[a[i][j]] = true;
      }
    }
  }
  for (int i = 1; i <= n; i++) {
    if (b[i])
      printf("%d\n", i);
  }
  return 0;
}
